smartsim/experiment.py

```diff
@@ -101,16 +115,17 @@ class Experiment:
     def __init__(
         self,
         name: str,
-        exp_path: t.Optional[str] = None,
-        launcher: str = "local",
+        exp_path: str | None = None,
+        *,  # Keyword arguments only
```

Comment:
It's mostly stylistic, but I added it for two main reasons:

The first is that there is sometimes a soft convention to make parameters that authors do not intend users to set often keyword only params so users do not accidentally set them by overrunning the positional args. That way an initialization like this:
```py
exp = Experiment("my-exp-name", "/path/to/my/exp",
        "some-positional-param-that-was-added-accidentally")
#       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#       Whoops!
```
is an error for experiment construction both at static analysis time and at runtime. Otherwise the error would not reveal itself at runtime until `exp.start_jobs` (to be renamed to `exp.start` in future) was called.

The second is for backward comparability. If, in future, we decide that we want to add a named parameter to the `__init__` and we expect users will want to set it often (or at least more often than `settings_dispatcher`), it might make sense for us to put it earlier in the arg list so that users can supply it positionally. BUT going from
```py
def __init__(self,
             name: str,
             exp_path: str,
             settings_dispatcher: Dispatcher = default_dispatcher) -> None: ...
```
to
```py
def __init__(self,
             name: str,
             exp_path: str,
             my_new_commonly_set_param: str = "sensible-default",
             settings_dispatcher: Dispatcher = default_dispatcher) -> None: ...
```
is an API break because a user could be constructing an experiment like so
```py
my_dispatcher = Dispatcher(...)
Experiment("name", "/path/to/exp", my_dispatcher)
```
but going from
```py
def __init__(self,
             name: str,
             exp_path: str,
             *,
             settings_dispatcher: Dispatcher = default_dispatcher) -> None: ...
```
to
```py
def __init__(self,
             name: str,
             exp_path: str,
             my_new_commonly_set_param: str = "sensible-default",
             *,
             settings_dispatcher: Dispatcher = default_dispatcher) -> None: ...
```
is not because the old signature required `settings_dispatcher` be supplied as a keyword argument already.

Reply:
In short, its 100% just style!! I would be more than willing to allow this parameter to be accepted positionally!

Reply:
The `dispatch` parameter has been moved from experiment construction to `start_jobs` where it needs to be a keyword only argument to allow for a variadic number of positional arguments. Marking resolved!